src/vision/status_quo/barbara_wishes_for_easy_runtime_switch.md

```diff
@@ -0,0 +1,66 @@
+# 😱 Status quo stories: Barbara wishes for an easy runtime switch
+
+## 🚧 Warning: Draft status 🚧
+
+This is a draft "status quo" story submitted as part of the brainstorming period. It is derived from
+real-life experiences of actual Rust users and is meant to reflect some of the challenges that Async
+Rust programmers face today.
+
+## The story
+
+[Barbara] has been working on an async codebase for the past 5 years. It is extremely mature and
+quite large (in the millions of lines of code). The product has traditionally been run on the server
+only, but now they want to move large parts to the edge. After some investigation, they realize that
+the open-source async runtime they have been using (tokio), while great for server workloads, is
+not very performant on the low-end edge devices.
+
+Tokio makes tradeoffs in memory for performance and while performance is still a concern on
+the low-end edge devices Barbara is targetting, memory consumption is even a hard constraint there.
+Additionally, the usage of a threadpool by tokio does not help either as the targetted edge
+devices have either a single CPU thread or only a few.
+
+They had never planned to run on edge devices and so it was assumed that they would always stick
+to tokio. Barbara looks into the how much work it would be to convert their codebase from
+tokio to a custom runtime tuned to their workload. To their disapointment, they discover many
+areas, where their choice of runtime permeates the code base:
+
+* They extensively use runtime-dependent helpers (e.g., a `select!` macro).
+* The timers, tasks, and async locks API is all tokio-specific.
+* Above all else, tokio uses a custom version of `AsyncRead` and `AsyncWrite` traits which
```

Comment:
I think another factor would be that many of their dependencies are tied to tokio, which probably links to other stories.

Reply:
You mean `AsyncRead` and `AsyncWrite`? The timers and async locks [aren't at least](https://github.com/rust-lang/wg-async-foundations/pull/176#discussion_r625670427).

Reply:
I meant e.g. `reqwest` or `hyper`

Reply:
ah right, yeah that'd make sense.